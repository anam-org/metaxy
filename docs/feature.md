## Feature Version

Every feature has a **feature version** - hash of its complete specification:
- Feature key
- Field definitions and code versions
- Dependencies (feature-level and field-level)

Being more pedantic, the feature version is calculated by hashing together its field versions and the feature key. The field version is calculated by hashing together the field code version, key, and versions of its dependencies.

Feature and field versions are static and deterministic. They are available as class methods:

```python
class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    fields=[
        FieldSpec(key=FieldKey(["frames"]), code_version=1),
        FieldSpec(key=FieldKey(["audio"]), code_version=1),
    ],
)):
    pass

# Get feature version
print(VideoProcessing.feature_version())  # "a3f8b2c1"

# Change code_version
class VideoProcessing(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "processing"]),
    fields=[
        FieldSpec(key=FieldKey(["frames"]), code_version=2),  # Changed!
        FieldSpec(key=FieldKey(["audio"]), code_version=1),
    ],
)):
    pass

print(VideoProcessing.feature_version())  # "d7e9f4a2" (different!)
```

Every metadata row includes a `feature_version` column:
```python
metadata = pl.DataFrame({
    "sample_id": [1, 2, 3],
    "data_version": [...],
    "feature_version": "a3f8b2c1",  # Auto-added by store
})
```

This enables:
- **Current vs historical**: Filter by `feature_version` to get specific versions
- **Migration tracking**: Know which rows need updating
- **Audit trail**: See what feature definition produced each row


## User-Defined Metadata Alignment

When feature definitions change, you may need custom logic to align metadata with upstream changes. Override `Feature.align_metadata_with_upstream`:

### Default Behavior (Inner Join)

```python
class MyFeature(Feature, spec=...):
    # Default behavior: inner join on sample_id
    # Only keeps samples present in ALL upstream features
    pass

# During migration:
# - Reads current metadata
# - Inner joins with upstream on sample_id
# - Only samples present in both current AND upstream survive
```

### Custom Alignment Examples

#### One-to-Many Mapping

```python
class VideoFrames(Feature, spec=FeatureSpec(
    key=FeatureKey(["video", "frames"]),
    deps=[FeatureDep(key=FeatureKey(["videos"]))],
    ...
)):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Each video produces 30 frames.

        When videos upstream change, regenerate all frame sample IDs.
        """
        video_samples = upstream_metadata["videos"]["sample_id"]

        # Generate frame sample IDs
        frames = []
        for video_id in video_samples:
            for frame_idx in range(30):
                frames.append({
                    "sample_id": f"{video_id}_frame_{frame_idx}",
                    "video_id": video_id,
                    "frame_idx": frame_idx,
                })

        return pl.DataFrame(frames)
```

#### Conditional Filtering

```python
class ProcessedVideos(Feature, spec=...):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Only process videos longer than 10 seconds."""

        videos = upstream_metadata["videos"]

        # Filter by duration
        valid_videos = videos.filter(pl.col("duration") > 10)

        # Keep custom columns from current_metadata if available
        if len(current_metadata) > 0:
            return current_metadata.join(
                valid_videos.select(pl.col("sample_id")),
                on="sample_id",
                how="inner",  # Only keep samples that pass filter
            )
        else:
            return valid_videos.select(["sample_id", "duration", "path"])
```

#### Outer Join (Keep All Samples)

```python
class MergedFeature(Feature, spec=...):
    @classmethod
    def align_metadata_with_upstream(
        cls,
        current_metadata: pl.DataFrame,
        upstream_metadata: dict[str, pl.DataFrame],
    ) -> pl.DataFrame:
        """Keep union of all upstream samples (outer join)."""

        all_samples = set()
        for upstream_df in upstream_metadata.values():
            all_samples.update(upstream_df["sample_id"].to_list())

        return pl.DataFrame({"sample_id": sorted(all_samples)})
```
