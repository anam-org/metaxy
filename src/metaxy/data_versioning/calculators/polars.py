"""Polars implementation of data version calculator."""

from collections.abc import Callable
from typing import TYPE_CHECKING, Any

import narwhals as nw
import polars as pl
import polars_hash as plh

from metaxy.data_versioning.calculators.base import DataVersionCalculator
from metaxy.data_versioning.hash_algorithms import HashAlgorithm
from metaxy.utils.hashing import truncate_struct_column

if TYPE_CHECKING:
    from metaxy.models.feature_spec import BaseFeatureSpec, IDColumns
    from metaxy.models.plan import FeaturePlan


class PolarsDataVersionCalculator(DataVersionCalculator):
    """Calculates data versions using polars-hash.

    Accepts Narwhals LazyFrames and converts internally to Polars for hashing.
    Supports all hash functions available in polars-hash plugin.
    Default is xxHash64 for cross-database compatibility.
    """

    # Map HashAlgorithm enum to polars-hash functions
    _HASH_FUNCTION_MAP: dict[HashAlgorithm, Callable[[pl.Expr], pl.Expr]] = {
        HashAlgorithm.XXHASH64: lambda expr: expr.nchash.xxhash64(),  # pyright: ignore[reportAttributeAccessIssue]
        HashAlgorithm.XXHASH32: lambda expr: expr.nchash.xxhash32(),  # pyright: ignore[reportAttributeAccessIssue]
        HashAlgorithm.WYHASH: lambda expr: expr.nchash.wyhash(),  # pyright: ignore[reportAttributeAccessIssue]
        HashAlgorithm.SHA256: lambda expr: expr.chash.sha2_256(),  # pyright: ignore[reportAttributeAccessIssue]
        HashAlgorithm.MD5: lambda expr: expr.nchash.md5(),  # pyright: ignore[reportAttributeAccessIssue]
    }

    @property
    def supported_algorithms(self) -> list[HashAlgorithm]:
        """All algorithms supported by polars-hash."""
        return list(self._HASH_FUNCTION_MAP.keys())

    @property
    def default_algorithm(self) -> HashAlgorithm:
        """xxHash64 - fast and cross-database compatible."""
        return HashAlgorithm.XXHASH64

    def calculate_data_versions(
        self,
        joined_upstream: nw.LazyFrame[Any],
        feature_spec: "BaseFeatureSpec[IDColumns]",
        feature_plan: "FeaturePlan",
        upstream_column_mapping: dict[str, str],
        hash_algorithm: HashAlgorithm | None = None,
    ) -> nw.LazyFrame[Any]:
        """Calculate data_version using polars-hash.

        Args:
            joined_upstream: Narwhals LazyFrame with upstream data joined
            feature_spec: Feature specification
            feature_plan: Feature plan
            upstream_column_mapping: Maps upstream key -> column name
            hash_algorithm: Hash to use (default: xxHash64)

        Returns:
            Narwhals LazyFrame with data_version column added
        """
        algo = hash_algorithm or self.default_algorithm

        if algo not in self.supported_algorithms:
            raise ValueError(
                f"Hash algorithm {algo} not supported by PolarsDataVersionCalculator. "
                f"Supported: {self.supported_algorithms}"
            )

        # Convert Narwhals LazyFrame to Polars LazyFrame
        # Must collect first (LazyFrame doesn't have to_polars, only DataFrame does)
        pl_lazy = joined_upstream.collect().to_polars().lazy()

        hash_fn = self._HASH_FUNCTION_MAP[algo]

        # Build hash expressions for each field
        field_exprs = {}

        for field in feature_spec.fields:
            field_key_str = (
                field.key.to_string()
                if hasattr(field.key, "to_string")
                else "_".join(field.key)
            )

            field_deps = feature_plan.field_dependencies.get(field.key, {})

            # Build hash components
            components = [
                pl.lit(field_key_str),
                pl.lit(str(field.code_version)),
            ]

            # Add upstream data versions in deterministic order
            for upstream_feature_key in sorted(field_deps.keys()):
                upstream_fields = field_deps[upstream_feature_key]
                upstream_key_str = (
                    upstream_feature_key.to_string()
                    if hasattr(upstream_feature_key, "to_string")
                    else "_".join(upstream_feature_key)
                )

                data_version_col_name = upstream_column_mapping.get(
                    upstream_key_str, "data_version"
                )

                for upstream_field in sorted(upstream_fields):
                    upstream_field_str = (
                        upstream_field.to_string()
                        if hasattr(upstream_field, "to_string")
                        else "_".join(upstream_field)
                    )

                    components.append(
                        pl.lit(f"{upstream_key_str}/{upstream_field_str}")
                    )
                    components.append(
                        pl.col(data_version_col_name).struct.field(upstream_field_str)
                    )

            # Concatenate and hash
            concat_expr = plh.concat_str(*components, separator="|")
            hashed = hash_fn(concat_expr).cast(pl.Utf8)

            field_exprs[field_key_str] = hashed

        # Create data_version struct
        data_version_expr = pl.struct(**field_exprs)  # type: ignore[call-overload]

        result_pl = pl_lazy.with_columns(data_version_expr.alias("data_version"))

        # Convert back to Narwhals LazyFrame and apply truncation
        result_nw = nw.from_native(result_pl, eager_only=False)

        # Use truncate_struct_column to apply hash truncation if configured
        # This needs to be eager for the truncation function
        result_eager = result_nw.collect()
        result_truncated = truncate_struct_column(result_eager, "data_version")

        # Convert back to lazy
        return nw.from_native(result_truncated.to_native().lazy(), eager_only=False)
