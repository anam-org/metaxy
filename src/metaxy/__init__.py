from metaxy.models.container import ContainerDep, ContainerSpec, SpecialContainerDep
from metaxy.models.feature import Feature
from metaxy.models.feature_spec import FeatureDep, FeatureDepMetadata, FeatureSpec
from metaxy.models.types import ContainerKey, FeatureKey

__all__ = [
    "Feature",
    "FeatureDep",
    "FeatureDepMetadata",
    "FeatureSpec",
    "ContainerDep",
    "ContainerSpec",
    "SpecialContainerDep",
    "FeatureKey",
    "ContainerKey",
]
