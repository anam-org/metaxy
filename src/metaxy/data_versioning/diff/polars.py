"""Polars implementation of metadata diff resolver."""

import polars as pl

from metaxy.data_versioning.diff.base import DiffResult, MetadataDiffResolver


class PolarsDiffResolver(MetadataDiffResolver):
    """Identifies changed rows using Polars operations.

    Strategy:
    - Categorizes changes into added, changed, and removed
    - Uses LEFT/RIGHT JOINs to identify each category
    - Materializes once and splits into three DataFrames (efficient)
    """

    def find_changes(
        self,
        target_versions: pl.LazyFrame,
        current_metadata: pl.LazyFrame | None,
    ) -> DiffResult:
        """Find all changes between target and current.

        Args:
            target_versions: LazyFrame with calculated data_versions
            current_metadata: LazyFrame with current metadata, or None

        Returns:
            DiffResult with three materialized DataFrames
        """
        # Select only sample_id and data_version from target_versions
        # (it may have intermediate joined columns from upstream)
        target_versions = target_versions.select(["sample_id", "data_version"])

        if current_metadata is None:
            # No existing metadata - all target rows are new, nothing removed/changed
            empty_df = pl.DataFrame({"sample_id": [], "data_version": []})
            return DiffResult(
                added=target_versions.collect(),
                changed=empty_df,
                removed=empty_df,
            )

        # Join target with current to categorize changes
        # Keep only sample_id and data_version from current for comparison
        current_comparison = current_metadata.select(
            ["sample_id", pl.col("data_version").alias("__current_data_version")]
        )

        # LEFT JOIN target with current
        compared = target_versions.join(
            current_comparison,
            on="sample_id",
            how="left",
        )

        # Build lazy queries for each category
        added_lazy = (
            compared.filter(pl.col("__current_data_version").is_null())
            .drop("__current_data_version")
            .select(["sample_id", "data_version"])
        )

        changed_lazy = (
            compared.filter(
                pl.col("__current_data_version").is_not_null()
                & (pl.col("data_version") != pl.col("__current_data_version"))
            )
            .drop("__current_data_version")
            .select(["sample_id", "data_version"])
        )

        removed_lazy = current_metadata.join(
            target_versions.select("sample_id"),
            on="sample_id",
            how="anti",
        ).select(["sample_id", "data_version"])

        # Collect all three at once (single execution for shared computations)
        added_df, changed_df, removed_df = pl.collect_all(
            [added_lazy, changed_lazy, removed_lazy]
        )

        return DiffResult(
            added=added_df,
            changed=changed_df,
            removed=removed_df,
        )
