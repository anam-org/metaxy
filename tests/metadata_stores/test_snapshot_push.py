"""Tests for feature graph snapshot push functionality (GitHub issue #86).

This module tests the new metadata-only change detection feature in
record_feature_graph_snapshot() that enables graph snapshot pushes
without triggering migrations.
"""

from __future__ import annotations

import polars as pl
from syrupy.assertion import SnapshotAssertion

from metaxy import (
    Feature,
    FeatureDep,
    FeatureKey,
    FeatureSpec,
    FieldKey,
    FieldSpec,
)
from metaxy.metadata_store.memory import InMemoryMetadataStore
from metaxy.models.feature import FeatureGraph
from metaxy.models.types import SnapshotPushResult


def test_snapshot_push_result_model():
    """Test SnapshotPushResult NamedTuple structure and backward compatibility."""
    result = SnapshotPushResult(
        snapshot_version="abc123",
        already_recorded=False,
        metadata_changed=False,
        features_with_spec_changes=[],
    )

    # Test fields
    assert result.snapshot_version == "abc123"
    assert result.already_recorded is False
    assert result.metadata_changed is False
    assert result.features_with_spec_changes == []

    # Test tuple unpacking for backward compatibility
    snapshot_version, already_recorded, metadata_changed, features = result
    assert snapshot_version == "abc123"
    assert already_recorded is False
    assert metadata_changed is False
    assert features == []


def test_record_snapshot_first_time():
    """Test recording a brand new snapshot (computational changes).

    Scenario 1: Empty store, first push.
    Expected: already_recorded=False, metadata_changed=False
    """
    graph = FeatureGraph()

    with graph.use():

        class VideoFiles(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["video", "files"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["path"]), code_version="1")],
            ),
        ):
            pass

        with InMemoryMetadataStore() as store:
            result = store.record_feature_graph_snapshot()

            # Verify result
            assert isinstance(result, SnapshotPushResult)
            assert result.already_recorded is False
            assert result.metadata_changed is False
            assert result.features_with_spec_changes == []
            assert result.snapshot_version == graph.snapshot_version

            # Verify data was written to feature_versions table
            from metaxy.metadata_store.system_tables import FEATURE_VERSIONS_KEY

            versions_lazy = store._read_metadata_native(FEATURE_VERSIONS_KEY)
            assert versions_lazy is not None
            versions_df = versions_lazy.collect().to_polars()

            assert versions_df.height == 1
            assert versions_df["feature_key"][0] == "video/files"
            assert versions_df["snapshot_version"][0] == graph.snapshot_version


def test_record_snapshot_metadata_only_changes():
    """Test recording metadata-only changes (rename in FeatureDep).

    Scenario 2: Snapshot exists, but feature_spec_version changed
    (without changing feature_version).

    Example: Adding rename={"old": "new"} to a FeatureDep
    This changes feature_spec_version but NOT feature_version.
    """
    from metaxy.metadata_store.system_tables import FEATURE_VERSIONS_KEY

    # Version 1: No rename
    graph_v1 = FeatureGraph()
    with graph_v1.use():

        class Upstream(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["upstream"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["value"]), code_version="1")],
            ),
        ):
            pass

        class Downstream(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["downstream"]),
                deps=[FeatureDep(key=FeatureKey(["upstream"]))],  # No rename yet
                fields=[FieldSpec(key=FieldKey(["result"]), code_version="1")],
            ),
        ):
            pass

        snapshot_v1 = graph_v1.snapshot_version
        feature_version_v1 = Downstream.feature_version()
        spec_version_v1 = Downstream.feature_spec_version()

        # First push
        with InMemoryMetadataStore() as store:
            result1 = store.record_feature_graph_snapshot()
            assert result1.already_recorded is False
            assert result1.metadata_changed is False

            # Verify initial state
            versions_lazy = store._read_metadata_native(FEATURE_VERSIONS_KEY)
            assert versions_lazy is not None
            versions_df = versions_lazy.collect().to_polars()
            assert versions_df.height == 2  # upstream + downstream

            # Version 2: Add rename (metadata-only change)
            # This changes spec_version but NOT feature_version or snapshot_version
            graph_v2 = FeatureGraph()
            with graph_v2.use():

                class Upstream2(
                    Feature,
                    spec=FeatureSpec(
                        key=FeatureKey(["upstream"]),
                        deps=None,
                        fields=[FieldSpec(key=FieldKey(["value"]), code_version="1")],
                    ),
                ):
                    pass

                class Downstream2(
                    Feature,
                    spec=FeatureSpec(
                        key=FeatureKey(["downstream"]),
                        deps=[
                            FeatureDep(
                                key=FeatureKey(["upstream"]),
                                rename={"value": "renamed_value"},  # Added rename
                            )
                        ],
                        fields=[FieldSpec(key=FieldKey(["result"]), code_version="1")],
                    ),
                ):
                    pass

                snapshot_v2 = graph_v2.snapshot_version
                feature_version_v2 = Downstream2.feature_version()
                spec_version_v2 = Downstream2.feature_spec_version()

                # Verify: snapshot_version and feature_version are SAME
                assert snapshot_v2 == snapshot_v1, (
                    "Snapshot version should be unchanged (metadata-only change)"
                )
                assert feature_version_v2 == feature_version_v1, (
                    "Feature version should be unchanged (metadata-only change)"
                )

                # But spec_version is DIFFERENT
                assert spec_version_v2 != spec_version_v1, (
                    "Spec version should change (metadata changed)"
                )

                # Second push - should detect metadata-only change
                result2 = store.record_feature_graph_snapshot()

                # Verify result
                assert result2.already_recorded is True
                assert result2.metadata_changed is True
                assert "downstream" in result2.features_with_spec_changes
                assert "upstream" not in result2.features_with_spec_changes
                assert result2.snapshot_version == snapshot_v2

                # Verify new rows were appended
                versions_lazy_after = store._read_metadata_native(FEATURE_VERSIONS_KEY)
                assert versions_lazy_after is not None
                versions_df_after = versions_lazy_after.collect().to_polars()
                # Old rows + new row for downstream (upstream unchanged)
                assert versions_df_after.height == 3

                # Verify only downstream has new row
                downstream_rows = versions_df_after.filter(
                    pl.col("feature_key") == "downstream"
                )
                assert downstream_rows.height == 2  # Two versions of downstream


def test_record_snapshot_no_changes():
    """Test recording when nothing changed.

    Scenario 3: Snapshot exists with identical code.
    Expected: already_recorded=True, metadata_changed=False
    """
    graph = FeatureGraph()

    with graph.use():

        class VideoFiles(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["video", "files"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["path"]), code_version="1")],
            ),
        ):
            pass

        with InMemoryMetadataStore() as store:
            # First push
            result1 = store.record_feature_graph_snapshot()
            assert result1.already_recorded is False

            # Second push - identical code
            result2 = store.record_feature_graph_snapshot()

            # Verify result
            assert result2.already_recorded is True
            assert result2.metadata_changed is False
            assert result2.features_with_spec_changes == []
            assert result2.snapshot_version == graph.snapshot_version

            # Verify no new rows appended
            from metaxy.metadata_store.system_tables import FEATURE_VERSIONS_KEY

            versions_lazy = store._read_metadata_native(FEATURE_VERSIONS_KEY)
            assert versions_lazy is not None
            versions_df = versions_lazy.collect().to_polars()
            assert versions_df.height == 1  # Still only 1 row


def test_record_snapshot_partial_metadata_changes():
    """Test metadata changes for SOME features (not all).

    Only changed features should appear in features_with_spec_changes.
    """
    from metaxy.metadata_store.system_tables import FEATURE_VERSIONS_KEY

    # Version 1: Three features
    graph_v1 = FeatureGraph()
    with graph_v1.use():

        class FeatureA(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["feature_a"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["value"]), code_version="1")],
            ),
        ):
            pass

        class FeatureB(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["feature_b"]),
                deps=[FeatureDep(key=FeatureKey(["feature_a"]))],
                fields=[FieldSpec(key=FieldKey(["result"]), code_version="1")],
            ),
        ):
            pass

        class FeatureC(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["feature_c"]),
                deps=[FeatureDep(key=FeatureKey(["feature_a"]))],
                fields=[FieldSpec(key=FieldKey(["output"]), code_version="1")],
            ),
        ):
            pass

        with InMemoryMetadataStore() as store:
            result1 = store.record_feature_graph_snapshot()
            assert result1.already_recorded is False

            # Version 2: Change metadata for FeatureB and FeatureC (not FeatureA)
            graph_v2 = FeatureGraph()
            with graph_v2.use():

                class FeatureA2(
                    Feature,
                    spec=FeatureSpec(
                        key=FeatureKey(["feature_a"]),
                        deps=None,
                        fields=[FieldSpec(key=FieldKey(["value"]), code_version="1")],
                    ),
                ):
                    pass

                class FeatureB2(
                    Feature,
                    spec=FeatureSpec(
                        key=FeatureKey(["feature_b"]),
                        deps=[
                            FeatureDep(
                                key=FeatureKey(["feature_a"]),
                                rename={"value": "renamed_b"},  # Changed
                            )
                        ],
                        fields=[FieldSpec(key=FieldKey(["result"]), code_version="1")],
                    ),
                ):
                    pass

                class FeatureC2(
                    Feature,
                    spec=FeatureSpec(
                        key=FeatureKey(["feature_c"]),
                        deps=[
                            FeatureDep(
                                key=FeatureKey(["feature_a"]),
                                columns=("value",),  # Changed
                            )
                        ],
                        fields=[FieldSpec(key=FieldKey(["output"]), code_version="1")],
                    ),
                ):
                    pass

                result2 = store.record_feature_graph_snapshot()

                # Verify: only B and C changed, not A
                assert result2.already_recorded is True
                assert result2.metadata_changed is True
                assert "feature_b" in result2.features_with_spec_changes
                assert "feature_c" in result2.features_with_spec_changes
                assert "feature_a" not in result2.features_with_spec_changes
                assert len(result2.features_with_spec_changes) == 2

                # Verify correct rows appended
                versions_lazy = store._read_metadata_native(FEATURE_VERSIONS_KEY)
                assert versions_lazy is not None
                versions_df = versions_lazy.collect().to_polars()
                # Original 3 + 2 new rows (B and C)
                assert versions_df.height == 5


def test_record_snapshot_append_only_behavior():
    """Test append-only behavior: old rows preserved, new rows added with same snapshot_version."""
    from metaxy.metadata_store.system_tables import FEATURE_VERSIONS_KEY

    # Start with a proper setup for metadata-only change
    graph_v1 = FeatureGraph()
    with graph_v1.use():

        class Upstream(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["upstream"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["value"]), code_version="1")],
            ),
        ):
            pass

        class MyFeature(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["my_feature"]),
                deps=[FeatureDep(key=FeatureKey(["upstream"]))],
                fields=[FieldSpec(key=FieldKey(["result"]), code_version="1")],
            ),
        ):
            pass

        snapshot_v1 = graph_v1.snapshot_version

        with InMemoryMetadataStore() as store:
            # Push v1
            result1 = store.record_feature_graph_snapshot()
            assert result1.snapshot_version == snapshot_v1

            versions_lazy_v1 = store._read_metadata_native(FEATURE_VERSIONS_KEY)
            assert versions_lazy_v1 is not None
            versions_df_v1 = versions_lazy_v1.collect().to_polars()
            assert versions_df_v1.height == 2  # upstream + my_feature

            my_feature_rows_v1 = versions_df_v1.filter(
                pl.col("feature_key") == "my_feature"
            )
            assert my_feature_rows_v1.height == 1
            timestamp_v1 = my_feature_rows_v1["recorded_at"][0]
            my_feature_rows_v1["feature_spec_version"][0]

            # Change metadata (rename in FeatureDep)
            graph_v2 = FeatureGraph()
            with graph_v2.use():

                class Upstream2(
                    Feature,
                    spec=FeatureSpec(
                        key=FeatureKey(["upstream"]),
                        deps=None,
                        fields=[FieldSpec(key=FieldKey(["value"]), code_version="1")],
                    ),
                ):
                    pass

                class MyFeature2(
                    Feature,
                    spec=FeatureSpec(
                        key=FeatureKey(["my_feature"]),
                        deps=[
                            FeatureDep(
                                key=FeatureKey(["upstream"]),
                                rename={"value": "new_name"},  # Metadata change
                            )
                        ],
                        fields=[FieldSpec(key=FieldKey(["result"]), code_version="1")],
                    ),
                ):
                    pass

                snapshot_v2 = graph_v2.snapshot_version
                assert snapshot_v2 == snapshot_v1, (
                    "Metadata-only change should keep same snapshot_version"
                )

                result2 = store.record_feature_graph_snapshot()
                assert result2.metadata_changed is True

                # Verify append-only: old rows still exist
                versions_lazy_v2 = store._read_metadata_native(FEATURE_VERSIONS_KEY)
                assert versions_lazy_v2 is not None
                versions_df_v2 = versions_lazy_v2.collect().to_polars()

                # Total rows: original 2 (upstream + my_feature) + 1 new (my_feature updated)
                assert versions_df_v2.height == 3

                # Find my_feature rows
                my_feature_rows = versions_df_v2.filter(
                    pl.col("feature_key") == "my_feature"
                )

                # Should have 2 rows: old + new
                assert my_feature_rows.height == 2

                # All rows have same snapshot_version
                assert my_feature_rows["snapshot_version"].unique().to_list() == [
                    snapshot_v1
                ]

                # But different spec_versions
                spec_versions = sorted(
                    my_feature_rows["feature_spec_version"].to_list()
                )
                assert len(spec_versions) == 2
                assert spec_versions[0] != spec_versions[1]

                # Timestamps are different (new row has later timestamp)
                timestamps = my_feature_rows.sort("recorded_at")[
                    "recorded_at"
                ].to_list()
                assert timestamps[0] == timestamp_v1  # Old row preserved
                assert timestamps[1] > timestamp_v1  # New row has later timestamp


def test_record_snapshot_computational_change():
    """Test computational change (code_version change) creates NEW snapshot.

    This is NOT a metadata-only change - it should return already_recorded=False.
    """
    graph_v1 = FeatureGraph()
    with graph_v1.use():

        class MyFeature(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["my_feature"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["value"]), code_version="1")],
            ),
        ):
            pass

        snapshot_v1 = graph_v1.snapshot_version

        with InMemoryMetadataStore() as store:
            result1 = store.record_feature_graph_snapshot()
            assert result1.snapshot_version == snapshot_v1

            # Change code_version (computational change)
            graph_v2 = FeatureGraph()
            with graph_v2.use():

                class MyFeature2(
                    Feature,
                    spec=FeatureSpec(
                        key=FeatureKey(["my_feature"]),
                        deps=None,
                        fields=[
                            FieldSpec(key=FieldKey(["value"]), code_version="2")
                        ],  # Changed
                    ),
                ):
                    pass

                snapshot_v2 = graph_v2.snapshot_version

                # Verify snapshot_version changed
                assert snapshot_v2 != snapshot_v1, (
                    "Computational change should create new snapshot_version"
                )

                result2 = store.record_feature_graph_snapshot()

                # Should be treated as NEW snapshot (not metadata-only)
                assert result2.already_recorded is False
                assert result2.metadata_changed is False
                assert result2.snapshot_version == snapshot_v2


def test_snapshot_push_result_snapshot_comparison(snapshot: SnapshotAssertion):
    """Test complete flow with snapshot for regression testing."""
    from metaxy.metadata_store.system_tables import FEATURE_VERSIONS_KEY

    # Create features and push multiple times
    results = []

    # Push 1: New snapshot
    graph_v1 = FeatureGraph()
    with graph_v1.use():

        class Upstream(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["upstream"]),
                deps=None,
                fields=[FieldSpec(key=FieldKey(["value"]), code_version="1")],
            ),
        ):
            pass

        class Downstream(
            Feature,
            spec=FeatureSpec(
                key=FeatureKey(["downstream"]),
                deps=[FeatureDep(key=FeatureKey(["upstream"]))],
                fields=[FieldSpec(key=FieldKey(["result"]), code_version="1")],
            ),
        ):
            pass

        with InMemoryMetadataStore() as store:
            result1 = store.record_feature_graph_snapshot()
            results.append(
                {
                    "push": 1,
                    "already_recorded": result1.already_recorded,
                    "metadata_changed": result1.metadata_changed,
                    "features_with_spec_changes": result1.features_with_spec_changes,
                }
            )

            # Push 2: No changes
            result2 = store.record_feature_graph_snapshot()
            results.append(
                {
                    "push": 2,
                    "already_recorded": result2.already_recorded,
                    "metadata_changed": result2.metadata_changed,
                    "features_with_spec_changes": result2.features_with_spec_changes,
                }
            )

            # Push 3: Metadata-only change
            graph_v2 = FeatureGraph()
            with graph_v2.use():

                class Upstream2(
                    Feature,
                    spec=FeatureSpec(
                        key=FeatureKey(["upstream"]),
                        deps=None,
                        fields=[FieldSpec(key=FieldKey(["value"]), code_version="1")],
                    ),
                ):
                    pass

                class Downstream2(
                    Feature,
                    spec=FeatureSpec(
                        key=FeatureKey(["downstream"]),
                        deps=[
                            FeatureDep(
                                key=FeatureKey(["upstream"]),
                                rename={"value": "renamed"},
                            )
                        ],
                        fields=[FieldSpec(key=FieldKey(["result"]), code_version="1")],
                    ),
                ):
                    pass

                result3 = store.record_feature_graph_snapshot()
                results.append(
                    {
                        "push": 3,
                        "already_recorded": result3.already_recorded,
                        "metadata_changed": result3.metadata_changed,
                        "features_with_spec_changes": result3.features_with_spec_changes,
                    }
                )

                # Get final feature_versions table state
                versions_lazy = store._read_metadata_native(FEATURE_VERSIONS_KEY)
                assert versions_lazy is not None
                versions_df = versions_lazy.collect().to_polars()

                # Convert to dict for snapshot (exclude timestamps for stability)
                versions_dict = versions_df.select(
                    [
                        "feature_key",
                        "feature_version",
                        "feature_spec_version",
                        "snapshot_version",
                    ]
                ).to_dicts()

    # Snapshot all results
    assert {
        "results": results,
        "final_versions_count": len(versions_dict),
        "versions": versions_dict,
    } == snapshot
