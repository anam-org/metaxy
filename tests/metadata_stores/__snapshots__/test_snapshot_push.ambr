# serializer version: 1
# name: test_snapshot_push_result_snapshot_comparison
  dict({
    'final_versions_count': 3,
    'results': list([
      dict({
        'already_recorded': False,
        'features_with_spec_changes': list([
        ]),
        'metadata_changed': False,
        'push': 1,
      }),
      dict({
        'already_recorded': True,
        'features_with_spec_changes': list([
        ]),
        'metadata_changed': False,
        'push': 2,
      }),
      dict({
        'already_recorded': True,
        'features_with_spec_changes': list([
          'downstream',
        ]),
        'metadata_changed': True,
        'push': 3,
      }),
    ]),
    'versions': list([
      dict({
        'feature_key': 'downstream',
        'feature_spec_version': 'e7711922f20c617291e3cb12c4f1e4ffec01cc4b1556b68fb58c2ec99658d2ed',
        'feature_version': '7b8e38a11f800fb93c18c0e04c4609d7b52ec12b62177ba8355531d363f01751',
        'snapshot_version': '1ab4ecd9d7af535b843dc22416dd99d5e430317c0a2bcd01fe83ffc2d7392da8',
      }),
      dict({
        'feature_key': 'upstream',
        'feature_spec_version': '00808d59808a90635b39b203e91f6d63529edc0e396bcb656c6612b264b29603',
        'feature_version': '8a2ffeab8da447095c5ee7a77e5635a1e16e7f3605021732f50f7002fa258398',
        'snapshot_version': '1ab4ecd9d7af535b843dc22416dd99d5e430317c0a2bcd01fe83ffc2d7392da8',
      }),
      dict({
        'feature_key': 'downstream',
        'feature_spec_version': 'c612638b8026fba25880d66a98a8a1f37103c789deda66c5c9f6bd5f7b5a5d85',
        'feature_version': '7b8e38a11f800fb93c18c0e04c4609d7b52ec12b62177ba8355531d363f01751',
        'snapshot_version': '1ab4ecd9d7af535b843dc22416dd99d5e430317c0a2bcd01fe83ffc2d7392da8',
      }),
    ]),
  })
# ---
